modules/nlu/src/backend/pipelines/intents/svm_classifier.ts

```diff
@@ -0,0 +1,240 @@
+      const l1Points: sdk.MLToolkit.SVM.DataPoint[] = []
+
+      for (const { name: intentName, tokens } of intents) {
+        for (const utteranceTokens of tokens) {
+          if (utteranceTokens.length) {
+            const l0Vec = await FTWordVecFeaturizer.getFeatures(this.language, utteranceTokens, l0Tfidf[context])
+            const l1Vec = await FTWordVecFeaturizer.getFeatures(
+              this.language,
+              utteranceTokens,
+              l1Tfidf[context][intentName]
+            )
+
+            l0Points.push({
+              label: context,
+              coordinates: [...l0Vec, utteranceTokens.length]
+            })
+
+            l1Points.push({
+              label: intentName,
+              coordinates: [...l1Vec, utteranceTokens.length]
+            })
+          }
+        }
+      }
+
+      const svm = new this.toolkit.SVM.Trainer()
+      await svm.train(l1Points, progress => debugTrain('SVM => progress for INT', { context, progress }))
+      const modelStr = svm.serialize()
+
+      models.push({
+        meta: { context, created_on: Date.now(), hash: modelHash, scope: 'bot', type: 'intent-l1' },
+        model: new Buffer(modelStr, 'utf8')
+      })
+    }
+
+    const svm = new this.toolkit.SVM.Trainer({ kernel: 'linear', classifier: 'C_SVC' })
+    await svm.train(l0Points, progress => debugTrain('SVM => progress for CTX %d', progress))
+    const ctxModelStr = svm.serialize()
+
+    models.push({
+      meta: { context: 'all', created_on: Date.now(), hash: modelHash, scope: 'bot', type: 'intent-l0' },
+      model: new Buffer(ctxModelStr, 'utf8')
+    })
+
+    models.push({
+      meta: { context: 'all', created_on: Date.now(), hash: modelHash, scope: 'bot', type: 'intent-tfidf' },
+      model: new Buffer(
+        JSON.stringify({ l0Tfidf: l0Tfidf['__avg__'], l1Tfidf: _.mapValues(l1Tfidf, x => x['__avg__']) }),
+        'utf8'
+      )
+    })
+
+    return models
+  }
+
+  private computeTfidf(
+    intentsWTokens: {
+      tokens: string[][]
+      name: string
+      utterances: string[]
+      filename: string
+      slots: sdk.NLU.SlotDefinition[]
+      contexts: string[]
+    }[]
+  ): { l1Tfidf: { [context: string]: TfidfOutput }; l0Tfidf: TfidfOutput } {
+    const allContexts = _.chain<sdk.NLU.IntentDefinition[]>(intentsWTokens)
+      .flatMap(x => x.contexts)
+      .uniq()
+      .value()
+
+    const l0TfidfInput: TfidfInput = {}
+    const l1Tfidf: {
+      [context: string]: TfidfOutput
+    } = {}
+    for (const context of allContexts) {
+      const intents = intentsWTokens.filter(x => x.contexts.includes(context))
+      l0TfidfInput[context] = _.flatten(_.flatten(intents.map(x => x.tokens)))
+      const l1Input: TfidfInput = {}
+      for (const { name, tokens } of intents) {
+        l1Input[name] = _.flatten(tokens)
+      }
+      l1Tfidf[context] = tfidf(l1Input)
+    }
+    const l0Tfidf: TfidfOutput = tfidf(l0TfidfInput)
+    return { l0Tfidf, l1Tfidf }
+  }
```

Comment:
the above block makes no sense to me, this is like the ultimate brace championship 🤣